crates/host/src/wasmbus/jetstream.rs

```diff
@@ -0,0 +1,303 @@
+
+        // Compute all new links that do not exist in the host map, which we'll use to
+        // publish to any running providers that are the source or target of the link.
+        // Computing this ahead of time is a tradeoff to hold only one lock at the cost of
+        // allocating an extra Vec. This may be a good place to optimize allocations.
+        let new_links = {
+            let all_links = self.links.read().await;
+            spec.links
+                .iter()
+                .filter(|spec_link| {
+                    // Retain only links that do not exist in the host map
+                    !all_links
+                        .iter()
+                        .filter_map(|(source_id, links)| {
+                            // Only consider links that are either the source or target of the new link
+                            if source_id == spec_link.source_id() || source_id == spec_link.target()
+                            {
+                                Some(links)
+                            } else {
+                                None
+                            }
+                        })
+                        .flatten()
+                        .any(|host_link| *spec_link == host_link)
+                })
+                .collect::<Vec<_>>()
+        };
+
+        {
+            // Acquire lock once in this block to avoid continually trying to acquire it.
+            let providers = self.providers.read().await;
+            // For every new link, if a provider is running on this host as the source or target,
+            // send the link to the provider for handling based on the xkey public key.
+            for link in new_links {
+                if let Some(provider) = providers.get(link.source_id()) {
+                    if let Err(e) = self.put_provider_link(provider, link).await {
+                        error!(?e, "failed to put provider link");
+                    }
+                }
+                if let Some(provider) = providers.get(link.target()) {
+                    if let Err(e) = self.put_provider_link(provider, link).await {
+                        error!(?e, "failed to put provider link");
+                    }
+                }
+            }
+        }
+
+        // If the component is already running, update the links
+        if let Some(component) = self.components.write().await.get(id) {
+            *component.handler.instance_links.write().await = component_import_links(&spec.links);
+            // NOTE(brooksmtownsend): We can consider updating the component if the image URL changes
+        };
+
+        // Insert the links into host map
+        self.links.write().await.insert(id.to_string(), spec.links);
+
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    pub(crate) async fn process_component_spec_delete(
+        &self,
+        id: impl AsRef<str>,
+        _value: impl AsRef<[u8]>,
```

Comment:
While we're here: Do we even need these two parameters?

Reply:
Removed while we were at it 😄